main.go

```diff
@@ -1,664 +1,24 @@
-
-			// check if we need to change the ticker interval
-			if departedCheckSeconds != newConfig.ApplicationSettings.DepartedCheckIntervalSeconds {
-				aggregateDepartedTicker.Stop()
-				departedCheckSeconds = newConfig.ApplicationSettings.DepartedCheckIntervalSeconds
-				aggregateDepartedTicker = time.NewTicker(time.Duration(departedCheckSeconds) * time.Second)
-				lc.Info(fmt.Sprintf("Changing aggregate departed check interval to %d seconds.", departedCheckSeconds))
-			}
-
-		case req := <-app.snapshotReqs:
-			data, err := json.Marshal(snapshot)
-			if err == nil {
-				_, err = req.w.Write(data) // only write if there was no error already
-			}
-			req.result <- err
-
-		case err := <-confErrCh:
-			lc.Error("Configuration error.", "error", err.Error())
-		}
-	}
-}
-
-func persistSnapshot(lc logger.LoggingClient, snapshot []inventory.StaticTag) {
-	lc.Debug("Persisting inventory snapshot.")
-	data, err := json.Marshal(snapshot)
-	if err != nil {
-		lc.Warn("Failed to marshal inventory snapshot.", "error", err.Error())
-		return
-	}
-
-	if err := ioutil.WriteFile(filepath.Join(cacheFolder, tagCacheFile), data, filePerm); err != nil {
-		lc.Warn("Failed to persist inventory snapshot.", "error", err.Error())
-		return
-	}
-	lc.Info("Persisted inventory snapshot.", "tags", len(snapshot))
-}
-
-// setDefaultBehavior sets the behavior associated with the default device group.
-func (app *inventoryApp) setDefaultBehavior(b llrp.Behavior) error {
-	app.devMu.Lock()
-	err := app.defaultGrp.SetBehavior(app.devService, b)
-	app.devMu.Unlock()
-	return err
-}
-
-// pushEventsToCoreData will send one or more Inventory Events as a single EdgeX Event with
-// an EdgeX Reading for each Inventory Event
-func (app *inventoryApp) pushEventsToCoreData(ctx context.Context, events []inventory.Event) error {
-	now := time.Now().UnixNano()
-	readings := make([]models.Reading, 0, len(events))
-
-	var errs []error
-	for _, event := range events {
-		payload, err := json.Marshal(event)
-		if err != nil {
-			errs = append(errs, errors.Wrap(err, "error marshalling event"))
-			continue
-		}
-
-		resourceName := resourceInventoryEvent + string(event.OfType())
-		app.edgexSdk.LoggingClient.Info("Sending Inventory Event.",
-			"type", resourceName, "payload", string(payload))
-
-		readings = append(readings, models.Reading{
-			Value:  string(payload),
-			Origin: now,
-			Device: eventDeviceName,
-			Name:   resourceName,
-		})
-	}
-
-	edgeXEvent := &models.Event{
-		Device:   eventDeviceName,
-		Origin:   now,
-		Readings: readings,
-	}
-
-	ctx, cancel := context.WithTimeout(ctx, coreDataPostTimeout)
-	defer cancel()
-
-	if _, err := app.edgexSdk.EdgexClients.EventClient.Add(ctx, edgeXEvent); err != nil {
-		errs = append(errs, errors.Wrap(err, "unable to push inventory event(s) to core-data"))
-	}
-
-	if errs != nil {
-		return llrp.MultiErr(errs)
+	if err := app.RunUntilCancelled(); err != nil {
+		fmt.Printf("App RunUntilCancelled failed: %v\n", err)
```

Comment:
Is that a design choice you would like us to make? Right now there are two or so conditions which the error is returned only and not logged. Would you prefer us to log it there and remove the logging here?

Reply:
Yea, prefer that so the error is logged using the Logging Client rather than using `Printf`

Reply:
Actually I was referring to:
> ... app.RunUntilCancelled() should log the error, so no need to print, just exit.

I took that to mean you meant:

```go
if err := app.RunUntilCancelled(); err != nil {
    os.exit(1)
}
```

There are some places where the `err` returned is not logged at all. Should I log them in the `RunUntilCancelled` function, or is it fine to log them here?

```go
if err := app.RunUntilCancelled(); err != nil {
    app.LoggingClient().Error("App RunUntilCancelled error occurred", "err", err.Error())
    os.exit(1)
}
```

Reply:
Yes I would like to see all errors logged. 
The approach we took with the new App Service template is the wrapper makes sure everything get logged and then just returns a status code for the exit rather than error.
https://github.com/edgexfoundry/app-functions-sdk-go/blob/master/app-service-template/main.go#L56
Or could return bool where true means continue and false means there was an error so exit. 
Point is the logging of errors is handled in the App where it has access to the logger clients.

Reply:
But if you want to stick with returning error that is fine, just want all errors logged.

Reply:
OK, then I think I will stick to logging it in `main.go`